File: 430. Flatten a Multilevel Doubly Linked List - Medium/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        curr, stack = head, [];
        while curr:
            if curr.child:
                if curr.next:
                    stack.append(curr.next);
                curr.next, curr.child.prev, curr.child = curr.child, curr, None;
            if not curr.next and stack:
                temp = stack.pop();
                temp.prev, curr.next = curr, temp;
            curr = curr.next;
        return head;

File: 430. Flatten a Multilevel Doubly Linked List - Medium/solution.cpp
static auto _ = [] () { ios_base::sync_with_stdio(false); cin.tie(nullptr); return 0; } ();

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head, Node* rest = nullptr) {
        if (!head) {
            return rest;
        }
        
        head->next = flatten(head->child, flatten(head->next, rest));
        if (head->next) {
            head->next->prev = head;
        }
        
        head->child = nullptr;
        return head;
    }
};